docs/design/Extension_Management.md

```diff
@@ -0,0 +1,151 @@
+# Extension Management
+
+Upgrade Assistant's extensibility provides the ability for users to extend it beyond its original design. As we've worked with developers making extensions, it's clear that there are some issues with discoverability and maintainability of custom extensions. Some of the major pain points:
+
+- **Version management:** Need to be able to specify which extension version a project should use
+- **Distribution** Need to be able easily get a specific version (or update existing ones) of an extension
+- **Consistency between runs** Should be able to register an extension so subsequent runs automatically use it
+
+Throughout this document, the term `workspace` will be used to refer to a solution or project that is being updated along with the state file Upgrade Assistant generates.
+
+## User Experience
+
+Extensions will now be something that are managed at a workspace level. This will allow users to have a consistent experience as well as be able to manage their extensions.
+
+Below are the commands to manage extensions:
+
+```
+upgrade-assistant extension list
+```
+
+This command will list current extensions applied to a project.
+
+```
+upgrade-assistant extension add --name [name] [--source [source]]
+```
+
+This command will allow users to add extensions to a project.
+
+```
+upgrade-assistant extension remove --name [name]
+```
+
+This command will remove an installed extension from a current workspace.
+
+```
+upgrade-assistant extension update [--name [name]]
+```
+
+This command will update all extensions, or a specific one if the name is given.
+
+```
+upgrade-assistant extension restore
+```
+
+This command will restore extensions when they are not available
+
+> This may be able to be done automatically, but initially, we'll have it be a separate command
+
+```
+upgrade-assistant extension feed list [--source [source]]
+```
+
+This command will allow someone to list extensions available on a feed
+
+### Extension Author Commands
+
+```
+upgrade-assistant extension feed publish --path [path] --source [source]
+```
+
+This command will allow extension authors to publish an extension.
+
+```
+upgrade-assistant extension feed remove --name [name] --source [source]
+```
+
+This command will allow extension authors to remove a published extension.
+
+```
+upgrade-assistant extension feed clean --source [source]
+```
+
+This command will allow extension authors to automatically clean up a feed (i.e. if a file has been removed from the storage, this will notify the author as well as remove it from the feed metadata).
+
+## Workspace state file
+
+Each workspace already contains a state file that tracks progress of the tool. This state file will be updated to store extension information. This state file should be persisted into source control, which will then ensure developers working on a project will get similar results.
```

Comment:
Maybe we need a transient state file (more of a cache) and then an upgrade assistant workspace file. Thoughts?

Reply:
Some properties makes sense to be persisted, like entry point. If someone wants to change it they can, but then users in a large project wouldn't all have to know what it is.

Reply:
I'm onboard with separate transient and persisted files. I think entry point may belong in transient, though, since a large solution may have multiple entry points being worked on by different people concurrently.

Reply:
Ok - I still think that entrypoint should be a stored value (too many people don't know what an entrypoint should even be), but I've updated the design to create a new workspace file that is expected to be persisted. This can then be updated later for other properties we want to be saved across runs.